### PR TITLE
Adding listIndexOf method to KeywordScanner

### PR DESCRIPTION
**Pull request by nixon124**
Adding the ability to search for occurrences of a pattern and return a list of positions in which the pattern starts.

### CODE REVIEW

File: src/main/java/emissary/util/search/ByteMatcher.java [hunk 1]
```diff
@@ -252,6 +335,32 @@ public int indexIgnoreCase(byte[] pattern, int beginIndex, int endIndex) {
...
+     * @param beginIndex start index
+     * @param endIndex the index to stop searching at, exclusive
+     *
+     * @return list of positions
+     */
+    public ArrayList<Integer> indexListIgnoreCase(byte[] pattern, int beginIndex, int endIndex) {
+
+        // Impossible to find under these conditions
+        if (mydata == null || beginIndex > (mydata.length - pattern.length) || endIndex > mydata.length) {
+            return EMPTYLIST;
```
Comment (fbruton) [inline src/main/java/emissary/util/search/ByteMatcher.java:352]: Update other occurrences as well.
```suggestion
    public List<Integer> indexListIgnoreCase(byte[] pattern, int beginIndex, int endIndex) {

        // Impossible to find under these conditions
        if (mydata == null || beginIndex > (mydata.length - pattern.length) || endIndex > mydata.length) {
            return Collections.emptyList();
```

File: src/main/java/emissary/util/search/ByteMatcher.java
Comment (drivenflywheel) [inline src/main/java/emissary/util/search/ByteMatcher.java:1]: Within this file and in KeywordScanner.java, let's move the new `listIndexOf` overloads to be consecutive, and after all the 'indexOf` overloads.  (i.e., don't break up the `indexOf` overloads by mirroring their order.) 

Same thing for the new `indexListIgnoreCase` methods in this file

File: src/main/java/emissary/util/search/ByteMatcher.java [hunk 2]
```diff
@@ -16,6 +17,7 @@ public class ByteMatcher {
     private KeywordScanner scanner = null;
 
     public static final int NOTFOUND = -1;
+    public static final ArrayList<Integer> EMPTYLIST = new ArrayList<>();
```
Comment (fbruton) [inline src/main/java/emissary/util/search/ByteMatcher.java:20]:
```suggestion
```

File: src/main/java/emissary/util/search/ByteMatcher.java
Comment (drivenflywheel) [inline src/main/java/emissary/util/search/ByteMatcher.java:1]: In this class and in `KeywordScanner`, let's change the return type of the new methods from `ArrayList<Integer>` to `List<Integer>`

File: src/test/java/emissary/util/search/ByteMatcherTest.java [hunk 3]
```diff
@@ -86,6 +91,37 @@ void testIndexOfBytes() {
         assertEquals(DATA.indexOf("fox"), this.b.indexOf("fox".getBytes()), "Match pos same as string");
     }
 
+    @Test
+    void testListIndexOf() {
+        // Case Sensitive
+        ArrayList<Integer> findTestCaseSensitive = new ArrayList<>();
+        Collections.addAll(findTestCaseSensitive, 10, 26, 47);
```
Comment (drivenflywheel) [inline src/test/java/emissary/util/search/ByteMatcherTest.java:98]:
```suggestion
        List<Integer> findTestCaseSensitive = List.of(10, 26, 47);
```

File: src/test/java/emissary/util/search/ByteMatcherTest.java [hunk 4]
```diff
@@ -86,6 +91,37 @@ void testIndexOfBytes() {
...
+        assertEquals(findTestCaseSensitive, this.bl.listIndexOf("test".getBytes(), 0, LIST_DATA.length()));
+
+        // String pattern param test
+        assertEquals(findTestCaseSensitive, this.bl.listIndexOf("test"));
+        assertEquals(findTestCaseSensitive, this.bl.listIndexOf("test", 0));
+        assertEquals(findTestCaseSensitive, this.bl.listIndexOf("test", 0, LIST_DATA.length()));
+
+        // Case-insensitive
+        ArrayList<Integer> findTestCaseInsensitive = new ArrayList<>();
+        Collections.addAll(findTestCaseInsensitive, 10, 26, 47, 53);
```
Comment (drivenflywheel) [inline src/test/java/emissary/util/search/ByteMatcherTest.java:111]:
```suggestion
        List<Integer> findTestCaseInsensitive = List.of(10, 26, 47, 53);
```

File: src/test/java/emissary/util/search/ByteMatcherTest.java [hunk 5]
```diff
@@ -86,6 +91,37 @@ void testIndexOfBytes() {
...
+        assertEquals(findTestCaseInsensitive, this.bl.indexListIgnoreCase("test".getBytes()));
+        assertEquals(findTestCaseInsensitive, this.bl.indexListIgnoreCase("test".getBytes(), 0));
+        assertEquals(findTestCaseInsensitive, this.bl.indexListIgnoreCase("test".getBytes(), 0, LIST_DATA.length()));
+
+        // String pattern param test
+        assertEquals(findTestCaseInsensitive, this.bl.indexListIgnoreCase("test"));
+        assertEquals(findTestCaseInsensitive, this.bl.indexListIgnoreCase("test", 0));
+        assertEquals(findTestCaseInsensitive, this.bl.indexListIgnoreCase("test", 0, LIST_DATA.length()));
+
+        String stop = "Stop";
```
Comment (drivenflywheel) [inline src/test/java/emissary/util/search/ByteMatcherTest.java:122]:
```suggestion
```

File: src/test/java/emissary/util/search/KeywordScannerTest.java [hunk 6]
```diff
@@ -14,12 +16,15 @@
 
 class KeywordScannerTest extends UnitTest {
     private static final byte[] DATA = "THIS is a test of the Emergency broadcasting system.".getBytes();
+    private static final byte[] LIST_DATA = "This is a test. Is this a test? Yes, this is a test.".getBytes();
     private KeywordScanner ks;
+    private KeywordScanner lks;
```
Comment (drivenflywheel) [inline src/test/java/emissary/util/search/KeywordScannerTest.java:21]: Make this a private variable in the `testListIndexOf` method

File: src/test/java/emissary/util/search/KeywordScannerTest.java [hunk 7]
```diff
@@ -96,6 +101,23 @@ void testCaseInsensitiveAllLUpperCasePatternFound() {
         assertEquals(10, this.ks.indexOf("TEST".getBytes()), "Case insensitive");
     }
 
+    @Test
+    void testListIndexOf() {
+        ArrayList<Integer> findTest = new ArrayList<>();
+        Collections.addAll(findTest, 10, 26, 47);
```
Comment (drivenflywheel) [inline src/test/java/emissary/util/search/KeywordScannerTest.java:107]:
```suggestion
        List<Integer> findTest = List.of(10, 26, 47);
```

File: src/test/java/emissary/util/search/KeywordScannerTest.java [hunk 8]
```diff
@@ -96,6 +101,23 @@ void testCaseInsensitiveAllLUpperCasePatternFound() {
...
+        assertEquals(findTest, this.lks.listIndexOf("test".getBytes()));
+        assertEquals(findTest, this.lks.listIndexOf("test".getBytes(), 0));
+        assertEquals(findTest, this.lks.listIndexOf("test".getBytes(), 0, LIST_DATA.length));
+
+        ArrayList<Integer> findIs = new ArrayList<>();
+        Collections.addAll(findIs, 2, 5, 21, 39, 42);
```
Comment (drivenflywheel) [inline src/test/java/emissary/util/search/KeywordScannerTest.java:113]:
```suggestion
        List<Integer> findIs = List.of(2, 5, 21, 39, 42);
```

**Review by drivenflywheel — CHANGES_REQUESTED** [review]

**Review by drivenflywheel — APPROVED** [review]

**Review by ldhardy — APPROVED** [review]

**Pull request merged**